Add base Trips Wizard

rachelekm opened the pull request:
## Overview

Implements the Trips wizard as well as the Add Trips modal. Adds logic to add/remove destinations to user profile buffer (using sample location data).

### Checklist

- [x] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following
      [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [x] Run `./scripts/format` to lint, format, and fix the application source code.
- [x] CI passes after rebase

### Demo


https://github.com/user-attachments/assets/c0e79dae-21db-42f8-a417-982f833a7c48


## Testing Instructions

 * `scripts/server`
 * Login as a user with no destinations
 * Move through the user profile wizard to get to the last step
 * Confirm last step displays no destinations and the Finish button is disabled
 * Click button to add a new trip
 * Confirm Add Trip modal opens
 * Click outside modal or on the back button
 * Confirm taken back to trips wizard step and no changes to wizard progress
 * Click to add trips again
 * Confirm "Add trip" button disabled with no purpose selection or no address input
 * Select purpose option (confirm styling matches figma)
 * Begin to type address in search suggestion and use button to clear selection (confirm this works)
 * The sample autocomplete suggestions list from Mapbox includes "Michigan", type an address input with this search term
 * Confirm autocomplete dropdown opens
 * Select option and confirm input populated with full selection text (I left an inline comment here, but I had trouble with keyboard-accessible focusing for button/dropdown. For the sake of time left it as is and can open a tech-debt card)
 * Confirm on selection that "Add trip" button no enabled
 * Select Add Trip and confirm guided back to trip wizard step
 * Confirm new destination now added to step
 * Confirm "Finish" now enabled
 * Select Finish and confirm user profile successfully updates with new data (Specifically, creates new Destination)
 * Confirm paths and times data successfully fetched for all destinations (under-the-hood this is creating a destination using one sample coordinate, so they will all be the same "place")
 * Confirm rankings calculate as expected
 * Logout and re-login, confirm no longer directed to Profile page


Resolves #720 #721

rachelekm committed:
Add initial trips step component

rachelekm committed:
Style add trip step content

rachelekm committed:
Create base add trip modal

rachelekm committed:
Add trip purpose selection

rachelekm committed:
Add async load Autocomplete base component

rachelekm committed:
Hook up Autocomplete to Add Trips

rachelekm committed:
Complete Add Trips Wizard logic

rachelekm committed:
Add Add Trips styling polish

rachelekm commented on src/frontend/src/components/base/Autocomplete/Autocomplete.tsx:
```diff
@@ -0,0 +1,106 @@
+        filteredSuggestions.items.length > 1 ||
+        (filteredSuggestions.items.length === 1 &&
+            filteredSuggestions.filterText !==
+                filteredSuggestions.items[0].name);
```
I ran into lots of trouble getting the Autocomplete menu to open/close as expected as well as navigate from the keyboard. Specifically, the menu of suggestions is always open, even when no search input or after a selection. This felt confusing from a user perspective. This component is a lot more constrained, so this is the best hack I could do that seemed to work without spending too much time on it for our budget constraints.

rachelekm replied:
I also could not get focus to work as expected, I don't have a fix for that here for the sake of time but will open a tech-debt card

aaronxsu replied:
Got it! With the remaining time, I think this is fine, re: focus and conditions to show the menu. I'd summarize your comments here and turn them into code comment.

aaronxsu commented on src/frontend/src/components/base/Autocomplete/Autocomplete.tsx:
```diff
@@ -0,0 +1,106 @@
+    const baseDropdownPopoverStyles = dropdownPopoverStyles();
+    const baseDropdownItemStyles = dropdownItemStyles();
+
+    let filteredSuggestions: AsyncListData<Suggestion> = useAsyncList({
```
This is an interesting first-use of `useAsyncList` in this project 👏🏼. I'd suggest adding some comments on what it does!

aaronxsu commented on src/frontend/src/pages/Discover/Profile/AddTripModal.tsx:
```diff
@@ -0,0 +1,224 @@
+        }
+    };
+
+    const purposesMap = Object.values(Place).map(value => ({
```
Would it make sense to move it out of the component as this won't change per component props?

aaronxsu commented on src/frontend/src/pages/Discover/Profile/StepTrips.tsx:
```diff
@@ -0,0 +1,97 @@
+                className="justify-normal text-gray-700 text-lg font-bold"
+                onPress={() => setIsAddTripModalOpen(true)}
+            >
+                {t("userTrip.wizard.stepAddTrip.addATrip")}
```
I'd consider conditionally display the copy here so that when there is no destination, it reads "Add a trip", and when there is more than 1, "Add another trip"

aaronxsu reviewed (commented):
This is looking great! I left some comments below. Additionally, when testing, I caught some edge cases that I'd like us to tackle:
1. As shown in the screenshot, the gap between the trip items seems larger than that in the design
2. The calculation of the primary destination seems a bit off as the two I added are both seen as primary. My suggestion: let's not assign primary in this modal step, but we take care of that in the finish button callback function: i.e. before submitting via the API, we mark the very first destination in the buffer as primary.
3. The `AddTripModal` state might need to be cleaned. For example: after I enter the first destination, click on "Add trip" button, it goes back to the main modal; then I click on "Add another trip", I am back to the add trip modal, but this time, the purpose is already selected (the same purpose as I selected in the previous trip) and the "Add trip" button is enabled (it should be disabled at this stage)
4. I'd suggest we update the highlight border color of the autocomplete text box. Currently it's yellow/orange. I'd probably use one of our theme colors or the blue color used when an element is highlighted.

What do you think?

<img width="1205" height="883" alt="Image" src="https://github.com/user-attachments/assets/ac1d2fc9-5e64-49c1-bf58-a2906a4c6085" />

rachelekm committed:
Fix and polish Add Trip modal logic

rachelekm commented:
Thanks, @aaronxsu! I fixed the above logic issues and added in that styling polish. This is all set for another look! Noting that I rebased with develop so you may need to force pull again.

aaronxsu reviewed (approved):
Thanks for addressing the comments and fixing the edge cases! This is now looking great and working very smoothly. Very nice work 👏🏼 

This is good to go after a changelog entry

rachelekm committed:
Update CHANGELOG
